iree/turbine/kernel/boo/fusion/subgraph.py

```diff
@@ -0,0 +1,181 @@
+from torch.fx.graph import Graph
+from torch.fx.graph_module import GraphModule
+from torch.fx.node import Node
+from torch._functorch.aot_autograd import _aot_export_function
+
+from .schema import FusionSchema, OpFusionSpec, Target
+from ..ops.graph import get_autograd_function
+from ....support.logging import aot_logger as logger
+
+
+def _log_graph_module(label: str, gm: GraphModule):
+    logger.debug("%s:\n%s", label, gm.print_readable(print_output=False))
+
+
+def get_subgraph_replacement(subgraph: GraphModule):
+    _log_graph_module("Extracted SubGraph Module", subgraph)
+    fake_args = tuple(
+        [n.meta.get("val") for n in subgraph.graph.nodes if n.op == "placeholder"]
+    )
+    joint_sg, metadata, in_spec, out_spec = _aot_export_function(
+        subgraph.forward,
+        fake_args,
+        decompositions=None,
+    )
+    # TODO: do some minimal validation on the results of the above function.
+    # in_spec, _kw_in_spec = in_spec.children_specs
+    _log_graph_module("AOT Joint FWD/BWD Subgraph Module", joint_sg)
+    fake_args_joint = tuple(
+        [n.meta.get("val") for n in joint_sg.graph.nodes if n.op == "placeholder"]
+    )
+
+    custom_op = get_autograd_function(
+        joint_sg, fake_args_joint, num_fwd_outputs=metadata.num_forward_returns
+    )
+
+    single_node_graph = fused_subgraph(
+        subgraph,
+        custom_op,
+        (n for n in subgraph.graph.nodes if n.op == "placeholder"),
+        num_outputs=metadata.num_forward_returns,
+    )
+    _log_graph_module("Replacement Subgraph", single_node_graph)
+    return single_node_graph
+
+
+def fused_subgraph(
+    root, target: Target, input_nodes: Sequence[Node], num_outputs: int
+) -> GraphModule:
+    g = Graph()
+    new_inputs = []
+    for n in input_nodes:
+        new_node = g.placeholder(n.name)
+        new_node.meta = n.meta
+        new_inputs.append(new_node)
+    fused_node = g.call_function(target, tuple(new_inputs))
+    outputs = []
+    for i in range(num_outputs):
+        outputs.append(g.call_method("__getitem__", args=(fused_node, i)))
+    g.output(result=tuple(outputs))
+    return GraphModule(root, g)
+
+
+def extract_fusion_subgraph_modules(
+    src_gm: GraphModule, fusion_schema: FusionSchema
+) -> Tuple[Sequence[GraphModule], Sequence[Dict[Node, Node]]]:
+    """Traverses src_gm nodes in order. When a node matches a root op in the fusion_schema,
+    A new subgraph is created with the root op in addition to any adjacent nodes matching the schema.
+
+    All subgraphs found this way are returned in a list.
+    The corresponding mappings from src_gm nodes -> subgraph nodes is also returned as a list.
+    """
+    subgraph_projections: List[Dict[Node, Node]] = []
+    subgraph_inclusions: List[Dict[Node, Node]] = []
+    subgraphs: List[GraphModule] = []
+    used_nodes: Set[Node] = set()
+    for root in src_gm.graph.nodes:
+        # if this node is included in any subgraph already, we can't use it.
```

Comment:
We probably want to warn or at least debug log that it is the case here. In future, we may also want a non-greedy approach to deciding which of the potentially overlapping fusions to implement.